modules/madsenseBidAdapter.js

```diff
@@ -0,0 +1,257 @@
+  },
+};
+
+function validateBidRequest(bid) {
+  return (
+    _validateParams(bid) &&
+    _validateBanner(bid) &&
+    _validateVideo(bid)
+  );
+}
+
+function getCompanyId(validBidRequests) {
+  let companyId = validBidRequests[0].params.company_id;
+
+  if (validBidRequests[0].params.test) {
+    logMessage('madsense: test mode');
+    companyId = 'test';
+  }
+
+  return companyId;
+}
+
+function buildEndpointUrl(companyId) {
+  return `${spec.ENDPOINT}?company_id=${companyId}`;
+}
+
+function parseServerResponse(serverResponse, bidRequest) {
+  return converter.fromORTB({
+    response: serverResponse.body,
+    request: bidRequest.data,
+  }).bids;
+}
+
+function filterValidBids(bids) {
+  return bids
+    .map((bid) => {
+      if (bid.mtype === 2 && bid.adm) {
+        if (!config.getConfig('cache.url')) {
+          bid.vastXml = bid.adm;
+          delete bid.adm;
+        } else {
+          logError('Prebid Cache is not configured (madSense)');
+          return null;
+        }
+      }
+      return bid;
+    })
+    .filter((bid) => bid !== null);
+}
+
+function hasBannerMediaType(bidRequest) {
+  return !!deepAccess(bidRequest, 'mediaTypes.banner');
+}
+
+function hasVideoMediaType(bidRequest) {
+  return !!deepAccess(bidRequest, 'mediaTypes.video');
+}
+
+function _validateParams(bidRequest) {
+  if (!bidRequest.params) {
+    return false;
+  }
+
+  if (bidRequest.params.test) {
+    return true;
+  }
+
+  if (!bidRequest.params.company_id) {
+    logError('company_id not declared (madSense)');
+    return false;
+  }
+
+  const mediaTypesExists = hasVideoMediaType(bidRequest) || hasBannerMediaType(bidRequest);
+  if (!mediaTypesExists) {
+    return false;
+  }
+
+  return true;
+}
+
+function _validateBanner(bidRequest) {
+  if (!hasBannerMediaType(bidRequest)) {
+    return true;
+  }
+  const banner = deepAccess(bidRequest, 'mediaTypes.banner');
+  if (!Array.isArray(banner.sizes)) {
+    return false;
+  }
+
+  return true;
+}
+
+function _validateVideo(bidRequest) {
+  if (!hasVideoMediaType(bidRequest)) {
+    return true;
+  }
+
+  const videoPlacement = deepAccess(bidRequest, 'mediaTypes.video', {});
+  const videoBidderParams = deepAccess(bidRequest, 'params.video', {});
+  const params = deepAccess(bidRequest, 'params', {});
+
+  if (params && params.test) {
+    return true;
+  }
+
+  const videoParams = {
+    ...videoPlacement,
+    ...videoBidderParams,
+  };
+
+  if (!Array.isArray(videoParams.mimes) || videoParams.mimes.length === 0) {
+    logError('Invalid MIME types (madSense)');
+    return false;
+  }
+
+  if (!Array.isArray(videoParams.protocols) || videoParams.protocols.length === 0) {
+    logError('Invalid protocols (madSense)');
+    return false;
+  }
+
+  if (!videoParams.context) {
+    logError('Context not declared (madSense)');
+    return false;
+  }
+
+  if (videoParams.context !== 'instream') {
+    logError('Only instream context is supported (madSense)');
```

Comment:
Above you prefer video media type if a unit is dual type. Please prefer banner if context is not in stream if you only support instrea

Also these can all be warnings instead of errors

Reply:
Also, optional chaining is generally preferred to deepAccess

Reply:
@patmmccann Addressed the feedback: banner is now preferred over instream where applicable, and swapped out deepAccess for optional chaining. Please let us know if any further adjustments are needed.